src/BuildLoop.cpp

```diff
@@ -265,8 +290,31 @@ static NodeBuildResult::Enum ExecuteNode(BuildQueue* queue, RuntimeNode* node, M
 
     NodeBuildResult::Enum runActionResult = RunAction(queue, thread_state, node, queue_lock);
 
-    if (runActionResult == NodeBuildResult::kRanSuccesfully && queue->m_Config.m_AttemptCacheWrites && IsNodeCacheableByLeafInputsAndCachingEnabled(queue,node))
+    // If we see any file not generated by the graph dated in the future we will always treat input signature as incorrect.
+    // Meaning for every build we will rebuild this node until filesystem mtime catches up with the file mtime date.
+    if (thereIsAtLeastOneInputFileDatedInTheFuture)
+    {
+        RuntimeNodeSetInputSignatureMightBeIncorrect(node);
+        return runActionResult;
+    }
+
+    // If signatures don't match, someone touched files on disk while we were checking input signature or action was executing.
```

Comment:
Should we print a warning in this case, since the build might be inconsistent?

Reply:
We coud but if a user updates a file while the build is running I think they are already aware it may have unforseen consequences to the running build.

Reply:
The important thing here is that we gurantee on the next run everything is consistent.

Reply:
I can imagine cases where it's non-obvious that files have been written during the build (eg periodic auto-save features in some text editors), but I do not have insight into how likely this is to affect our users.

There might also be situations where a misconfigured build does bad things like moving or overwriting files multiple times- if we present these warnings to users and they're seen at unexpected times then these problems are more likely to be fixed.

Reply:
Yeah, for this one I think we can put a warning. It should only happen very rarley and as you say if users are running multiple build systems in parallel it may help catch bugs.

Reply:
`Log(kWarning, "concurrent modification of inputs detected while executing `%s`.", node->m_DagNode->m_Annotation.Get());`

works?